storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java

```diff
@@ -104,6 +106,22 @@ public class RemoteLogMetadataCache {
     // https://issues.apache.org/jira/browse/KAFKA-12641
     protected final ConcurrentMap<Integer, RemoteLogLeaderEpochState> leaderEpochEntries = new ConcurrentHashMap<>();
 
+    private final CountDownLatch initializedLatch = new CountDownLatch(1);
+
+    public void markInitialized() {
+        initializedLatch.countDown();
+    }
+
+    public void ensureInitialized() throws InterruptedException {
+        if (!initializedLatch.await(2, TimeUnit.MINUTES)) {
```

Comment:
I can see we will block for operations like `findOffsetByTimestamp` (invoked by clients), and the client request timeout default is 30 secs. I think the 2 mins wait is not appropriate here. Maybe 10 secs? Or any reason we chose 2 mins?

Reply:
In large clusters, we saw that it takes more than a minute to initialize. That's the reason we kept it 2 mins. We could also make this configurable. Thoughts? 
cc @kamalcph

Reply:
Both `FETCH` and `LIST_OFFSETS` APIs will hit this method. If the partition is not initialised within the timeout (2 mins / 10 secs), then we throw `RemoteResourceNotFoundException` which will in turn throw `UNKNOWN_SERVER_ERROR` back to the caller. 

Consumer will retry the `FETCH` request as `UNKNOWN_SERVER_ERROR` is retriable. The `LIST_OFFSETS` call can originate from both Consumer and Admin client. Consumer retries the request on receiving retriable error but `Admin` client won't. If the user tries to increase the request timeout from 30 secs to 2 min in AdminClient, then the `LIST_OFFSETS` request should succeed.

This timeout specified here is kind of upper boundary where we expect the RLMM to initialise the partition within that amount of time. 

If the request timeout is configured to default value of 30 secs on the client, then the request gets timed out and the client retries the request. In case of `AdminClient`, the user gets to know that the request was failed due to TimeoutException and will increase the request timeout as appropriate.

Reply:
We can reduce the timeout to 10 seconds but should think of a solution for Admin#listOffsets call.

Reply:
Thanks for the explanation, @kamalcph ! 

One correction: `UNKNOWN_SERVER_ERROR` is **unretriable**. So if we throw `RemoteResourceNotFoundException`, the client won't be retried. 

I think we don't have to consider if client will auto retry or not when facing retriable error. IMO, sending a request and returned a request timeout error without any explanation (suppose the 2 min timeout is not expired here), which is not good. 

I think reduce to a small number (I think 10 secs is still too long, maybe 1 sec, or return error directly if it's not initialized), and wrap the `RemoteResourceNotFoundException` with a retriable error, ex: `ReplicaNotAvailableException`, and add clear messages there. I think as long as we returned clear error reason with a retriable exception, this is already a good API. We don't want to block the request for a long time because we're waiting for some state change. Like `COORDINATOR_LOAD_IN_PROGRESS` error, it's clear that we returned this error and want the client retry later, instead of waiting in the broker for coordinator load completion. Does that make sense?

Reply:
The java client retries when it receives `UNKNOWN_SERVER_ERROR` not sure for other clients. As per the protocol contract, the clients are not expected to retry when it receives this error.

We also have one more API `isInitialized` that can be used instead of `ensureInitialized` if we want to unblock the caller immediately. This latch will only block the remote-log-reader and remote-log-deleter threads until the partition gets initialised and won't block the request handler threads.

The idea to increase the timeout upto 2 minutes was similar to delayed purgatory where the requests are hold until it gets served within the request timeout.

Reply:
Currently, the LIST_OFFSETS call is sync which is handled by the request handler threads so the rquest-handler-thread will be blocked here. Until KAFKA-13560 gets implemented, it's good to reduce the timeout to 1s (or) throw error immediately as suggested.

Reply:
@kamalcph , thanks for the update!
@abhijeetk88 , could you reduce the timeout to 1s (or) throw error immediately as discussed above? Thanks.

Reply:
@showuon I have addressed this. Added a new retryable exception. Let me know if we need to rename it.